File: root2leafsum.java
// T.C : O(N)
// S.C : O(H)

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int sum;
    public int sumNumbers(TreeNode root) {
        if(root == null) return sum;
         
        helper(root,0);
        return sum;
    }
    
    private void helper(TreeNode root,int num){
        // base
        if(root == null) return;
        // logic
        
        helper(root.left, num*10+root.val);
        // check if on leaf node
        // sum for left part
        if(root.left==null && root.right == null){
            sum = sum + num*10 + root.val;
            return;
        }
         // sum for right part
        helper(root.right, num*10+root.val);
    }
}

File: Root2LeafSum.py
# Time Complexity : O(N)
# Space Complexity : O(H) ~> Recursion Stack
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


# Your code here along with comments explaining your approach

'''Top Down Recursion with Global Result'''
class Solution:
    def __init__(self):
        self.result = 0
        
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        def recur(root, number):
            if root.left == None and root.right == None:
                curr = number * 10 + root.val
                self.result += curr
                return 
                       
            curr = number * 10 + root.val
            if root.left:
                recur(root.left, curr)
            
            if root.right:
                recur(root.right, curr)
        
        
        recur(root, 0)
        return self.result

'''Bottom Up Recursion'''

class Solution:
    
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        
        def recur(root, number):
            if root.left == None and root.right == None:
                curr = number * 10 + root.val
                return curr
                       
            curr = number * 10 + root.val
            left = 0
            right = 0
            if root.left:
                left = recur(root.left, curr)
            
            if root.right:
                right = recur(root.right, curr)
            
            return left + right
        
        return recur(root, 0)